Change NXsample in NXmpes using the new base classes, introduce NXactuator

lukaspie opened the pull request:
- Sample preparation goes to NXsample_history
- Temperature, bias, and gas pressure go to NXenvironment -> this could also be on instrument (or on both with a link)
- chemical_formula goes to NXsubstance

@domna @rettigl @RubelMozumder

domna commented:
Thank you! The PR removes and alters a lot of fields which have nothing to do with NXsample. Is this a problem with the new Makefile?

lukaspie commented:
It shouldn't, I used the old Makefile on mpes-refactor. It seems that the removed fields are those that are on the mpes-refactor branch, but not yet on FAIRmat, so something must have gone wrong. I will have a look again.

domna commented on contributed_definitions/nyaml/NXmpes.yaml:
```diff
@@ -1,28 +1,18 @@
     start_time(NX_DATE_TIME):
       doc: |
         Datetime of the start of the measurement.
-    end_time(NX_DATE_TIME):
```
As you can see here the changes are also reflected in the `yaml` file. So somehow an old version slipped into this branch.

lukaspie replied:
Fixed, now using the correct yaml version in mpes-refactor.

lukaspie commented:
The PR is now ready to be reviewed. 

One thing that I noticed is that when using NXenvironment in NXsample instead of simple NX_FLOAT (e.g. for temperature, bias), these should contain a link to a corresponding entity in NXmanipulator. However, NXmanipulator does contain these fields as NX_FLOAT as well (sample_temperature, sample_bias, ...). Therefore, NXmanipulator needs to be changed when we refactor NXinstrument. For now, I have kept the fields in NXmanipulator unchanged.

domna commented on contributed_definitions/nyaml/NXmpes.yaml:
```diff
@@ -230,58 +232,139 @@ NXmpes(NXobject):
+          exists: recommended
+          doc: |
+            The chemical formula of the sample.
       atom_types:
```
Is this correct here or should it go to NXsubstance?

lukaspie replied:
We don't have atom_types in NXsubstance. atom_types is supposed to contain all elements in the sample (incl. subcomponents), while NXsubstance is just for a single, undividable chemical substance.

lukaspie commented:
Yes, this is correct and it also works like it is. My point is that it would be nice to store these parameters as an NXenvironment on the manipulator side as well. This would allow us to have the description of the NXsensors measuring the temperature, pressure, and so on, within NXinstrument (where they should arguably belong).

domna commented on contributed_definitions/nyaml/NXmpes.yaml:
```diff
@@ -230,58 +232,139 @@ NXmpes(NXobject):
+              Details about the method of sample preparation before the MPES experiment.
+          notes:
+            exists: optional
+      temperature(NXenvironment):
```
Ok, now I understand your point with the temperature. Yes we should change the docstring in NXmanipulator, that it is the same concept and value as NXsample/thermocouple/value

domna replied:
> Yes, this is correct and it also works like it is. My point is that it would be nice to store these parameters as an NXenvironment on the manipulator side as well. This would allow us to have the description of the NXsensors measuring the temperature, pressure, and so on, within NXinstrument (where they should arguably belong).

Yes, now I understood what you mean. I like the suggestion to add and link the NXsensor in the manipulator as this fits in the instrument part of the experiment.

domna reviewed (approved):
Looks good, except that I'm not sure if `atom_types` is missing one indentation and should go under `NXsubstance`.

rettigl commented on contributed_definitions/nyaml/NXactuator.yaml:
```diff
@@ -0,0 +1,113 @@
+      Short name of actuator used e.g. on monitor display program
+  attached_to:
+    doc: |
+      where actuator is attached to ("sample" | "can")
```
see above

rettigl commented on contributed_definitions/nyaml/NXactuator.yaml:
```diff
@@ -0,0 +1,113 @@
+  attached_to:
+    doc: |
+      where actuator is attached to ("sample" | "can")
+  geometry(NXgeometry):
+    deprecated: |
+      Use the field `depends_on` and :ref:`NXtransformations` to position the beamstop and NXoff_geometry to describe its shape instead
```
see above

rettigl commented on contributed_definitions/nyaml/NXactuator.yaml:
```diff
@@ -0,0 +1,113 @@
+  actuation:
+    doc: |
+      name for measured signal
+    enumeration: [temperature, pH, magnetic_field, electric_field, current, conductivity, resistance, voltage, pressure, flow, stress, strain, shear, surface_pressure]
```
see above

rettigl commented on contributed_definitions/nyaml/NXmanipulator.yaml:
```diff
@@ -0,0 +1,64 @@
+  sample_heater(NXactuator):
+    doc: |
+      Device for applying heating power for temperature control.
+  sample_thermocouple(NXsensor):
+    doc: |
```
temperature_sensor

Often, other sensors than thermocouples are used (e.g. Silicon diodes)

lukaspie replied:
Agreed

rettigl commented on contributed_definitions/nyaml/NXmpes.yaml:
```diff
@@ -176,16 +176,113 @@ NXmpes(NXobject):
-          Manipulator for positioning of the sample.
-        sample_temperature(NX_FLOAT):
+          Manipulator for positioning of the sample.      
+        sample_thermocouple(NXsensor):
```
temperature_sensor

rettigl commented on contributed_definitions/nyaml/NXmpes.yaml:
```diff
@@ -176,16 +176,113 @@ NXmpes(NXobject):
           exists: recommended
-          unit: NX_TEMPERATURE
-        drain_current(NX_FLOAT):
+          doc: Thermocouple measuring the sample temperature.
```
Temperature sensor

rettigl commented on contributed_definitions/nyaml/NXmpes.yaml:
```diff
@@ -176,16 +176,113 @@ NXmpes(NXobject):
+            enumeration: [temperature]
+          type:
+            exists: recommended
+          value(NX_FLOAT):
+            exists: recommended
+            unit: NX_TEMPERATURE
+            doc: |
+              In the case of a fixed temperature measurement, this is the scalar temperature measured
+              by the sample thermocouple.
+          value_log(NXlog):
+            exists: optional
+            unit: NX_TEMPERATURE
+            doc: |
+              In the case of an experiment in which the temperature is changed and recorded,
+              this is an array of length m of temperatures.
```
This means one could provide this group, but none of the two options...

rettigl commented on contributed_definitions/nyaml/NXmpes.yaml:
```diff
@@ -176,16 +176,113 @@ NXmpes(NXobject):
+            exists: optional
+            unit: NX_TEMPERATURE
+            doc: |
+              In the case of an experiment in which the temperature is changed and recorded,
+              this is an array of length m of temperatures.
```
temperature setpoints

lukaspie replied:
👍

rettigl commented on contributed_definitions/nyaml/NXmpes.yaml:
```diff
@@ -230,58 +327,90 @@ NXmpes(NXobject):
-          A descriptor to keep track of the treatment of the sample before entering the
-          photoemission experiment. Ideally, a full report of the previous operations, in
-          any format (NXnote allows to add pictures, audio, movies). Alternatively, a
-          reference to the location or a unique identifier or other metadata file. In the
-          case these are not available, free-text description.
+          For samples containing a single pure substances. For mixtures use the
+          NXsample_component_set and NXsample_component group in NXsample instead.
```
NXsubstance is required. How does this work with NXsample_component_set then? Not clear how to use it, and what fields to provide.

lukaspie replied:
You are right, I thought in the existing NXmpes, `chemical_formula` was required. Since it is only recommended, we can also recommend NXsubstance.

rettigl commented on contributed_definitions/nyaml/NXmpes.yaml:
```diff
@@ -230,58 +327,90 @@ NXmpes(NXobject):
-      temperature(NX_FLOAT):
-        unit: NX_TEMPERATURE
+          Sample temperature (either controlled or just measured).
+        thermocouple(NXsensor):
```
temperature_sensor

rettigl reviewed (commented):
This goes in a good direction, I've made some comments where I see room fro improvements or don't understand things. I think it would be important to have a working example implementing these changes into a nexus file alongside (you can use the mpes WSe2 example e.g.),

lukaspie commented:
Thanks for your feedback @rettigl! The diff may make it seem that I introduced NXsensor by myself, but it was already there and designed for NXsts and/or NXiv_temp by @RubelMozumder and @sanbrock, I believe (see also [NXsensor_scan](https://fairmat-nfdi.github.io/nexus-fairmat-proposal/9636feecb79bb32b828b1a9804269573256d7696/classes/contributed_definitions/NXsensor_scan.html#nxsensor-scan)). I agree that some of the fields that are in NXsensor are somewhat unintuitive. As an example, the difference between `value` and `value_log` are confusing. I suppose that NXsensor was originally intended to be used as an actuator/sensor hybrid, i.e., the field `value` was meant to store setpoints and `value_log` was to be used for the measured values. As a consequence of NXsample already existing, NXactuator as of now was simply designed to mimic NXsensor.

My suggestion would be to change both NXsensor and NXactuator such that the fields and docstrings are clearer. However, this can only be done together with the original NXsensor developers, such that their appdefs are not broken.

Since you were also asking for an example, I have started to change the ELN for the XPS example [here](https://gitlab.mpcdf.mpg.de/nomad-lab/nomad-remote-tools-hub/-/blob/xps_eln_update/docker/xps/example/Example%202%20-%20SPECS%20native%20SLE%20data%20format/eln_data_sle.yaml). This is however also still very much in progress and depends on the changes made here.

rettigl commented:
> Thanks for your feedback @rettigl! The diff may make it seem that I introduced NXsensor by myself, but it was already there and designed for NXsts and/or NXiv_temp by @RubelMozumder and @sanbrock, I believe (see also [NXsensor_scan](https://fairmat-nfdi.github.io/nexus-fairmat-proposal/9636feecb79bb32b828b1a9804269573256d7696/classes/contributed_definitions/NXsensor_scan.html#nxsensor-scan)). I agree that some of the fields that are in NXsensor are somewhat unintuitive. As an example, the difference between `value` and `value_log` are confusing. I suppose that NXsensor was originally intended to be used as an actuator/sensor hybrid, i.e., the field `value` was meant to store setpoints and `value_log` was to be used for the measured values. As a consequence of NXsample already existing, NXactuator as of now was simply designed to mimic NXsensor.

I understood that this was not brewed up by you alone, I still saw it a good opportunity to comment on these points in this context. It's probably a good idea to discuss these issues again soon in one of our meetings.

lukaspie commented:
I have merged here the changes from the `mpes-refactor` branch. I have also changed the value log in the groups using `NXsensor`.

lukaspie commented on contributed_definitions/nyaml/NXmpes.yaml:
```diff
@@ -288,34 +288,129 @@ NXmpes(NXobject):
-          unit: NX_TEMPERATURE
-        drain_current(NX_FLOAT):
+          doc: Temperature sensor measuring the sample temperature.
+          name:
+            exists: recommended
+          measurement:
+            enumeration: [temperature]
+          type:
+            exists: recommended
+          value(NX_FLOAT):
+            exists: recommended
+            unit: NX_TEMPERATURE
+            doc: |
+              In the case of a fixed temperature measurement, this is the scalar temperature measured
+              by the sample temperature sensor.
+          value_log(NXlog):
+            exists: optional
+            value(NX_NUMBER):
+              doc: |
+                In the case of an experiment in which the temperature is changed and recorded,
+                this is an array of length m of temperatures.
+              unit: NX_TEMPERATURE
+        sample_heater(NXactuator):
           exists: recommended
-          unit: NX_CURRENT
-        sample_bias(NX_FLOAT):
+          doc: Device to heat the sample.
+          name:
+            exists: recommended
+          actuation:
+            enumeration: [temperature]
+          type:
+            exists: recommended
+          setpoint_log(NXlog):
+            exists: recommended
+            value:
+              unit: NX_TEMPERATURE
+          output_log(NXlog):
+            value:
+              unit: NX_TEMPERATURE
```
I think the main blocker with this PR is that `NXsensor` is somewhat poorly designed for this use case, especially the distinction between `value` and `value_log`. This leads to the problems outlined above by @rettigl , i.e., deprecated fields, unnecessary limitations, etc. Ideally, `NXsensor` should be restructured to be in line with `NXactuator`. You can see the difference in using the two base classes here. However, since `NXsensor` is an existing base class and shall not be changed to not break other NeXus use cases, we need to use the workarounds in ´NXmpes/NXmanipulator´.

lukaspie commented:
Ready for review @domna @rettigl

rettigl commented on contributed_definitions/nyaml/NXactuator.yaml:
```diff
@@ -0,0 +1,106 @@
+      Can be a scalar or a vector (of [n] actuations).
+    dimensions:
+      dim: [[1, n]]
+  setpoint_deriv1(NX_FLOAT):
+    doc: |
+      Nominal/average first derivative of setpoint
+      e.g. strain rate
+      - same dimensions as "setpoint" (may be a vector)
+    dimensions:
+      dim: [[1, ]]
+      dim_parameters:
+        ref: ['setpoint']
+  setpoint_deriv2(NX_FLOAT):
+    doc: |
+      Nominal/average second derivative of setpoint
+      - same dimensions as "setpoint" (may be a vector)
+    dimensions:
+      dim: [[1, ]]
+      dim_parameters:
+        ref: ['setpoint']
```
What are these good for?
For the sensor data, the derivatives (and integrals) are used for PID control, but for the setpoints I don't see any obvious use case.

lukaspie replied:
You are right -> removed

rettigl commented on contributed_definitions/nyaml/NXactuator.yaml:
```diff
@@ -0,0 +1,106 @@
+  (NXpid):
+    doc: |
+      If the actuator is PID-controlled, the settings of the PID controller can be stored here.
+  (NXoff_geometry):
+    exists: ['min', '0']
+    doc: |
+      This group describes the shape of the actuator when necessary.
```
How is that supposed to be used?
I suggest to not merely reproduce all elements in NXsensor, but only those we think are useful.

rettigl commented on contributed_definitions/nyaml/NXactuator.yaml:
```diff
@@ -0,0 +1,106 @@
+    exists: ['min', '0']
+    doc: |
+      This group describes the shape of the actuator when necessary.
+  \@default:
+    doc: |
+      .. index:: plotting
+      
+      Declares which child group contains a path leading
+      to a :ref:`NXdata` group.
+      
+      It is recommended (as of NIAC2014) to use this attribute
+      to help define the path to the default dataset to be plotted.
+      See https://www.nexusformat.org/2014_How_to_find_default_data.html
+      for a summary of the discussion.
```
There is no NXdata child group defined here, so I don't see why this should be here.

rettigl replied:
Seems to me to also be an error in NXsensor

lukaspie replied:
Only removed in NXactuator for now (in order not to overwrite NIAC's NXsensor) -> should be discussed with them at some point

rettigl commented on contributed_definitions/nyaml/NXactuator.yaml:
```diff
@@ -0,0 +1,106 @@
+  setpoint_log(NXlog):
+    doc: |
+      Time history of actuator setpoints.
+  setpoint_deriv1_log(NXlog):
+    doc: |
+      Time history of first derivative of actuator setpoints.
+  setpoint_deriv2_log(NXlog):
+    doc: |
+      Time history of second derivative of actuator setpoints.
```
same

lukaspie replied:
removed

rettigl commented on contributed_definitions/nyaml/NXactuator.yaml:
```diff
@@ -0,0 +1,106 @@
+      This is the group recommended for holding the chain of translation
+      and rotation operations necessary to position the component within
+      the instrument. The dependency chain may however traverse similar groups in
+      other component groups.
```
NXfabrication missing

rettigl commented on contributed_definitions/nyaml/NXmanipulator.yaml:
```diff
@@ -13,36 +13,106 @@ NXmanipulator(NXobject):
+    measurement:
+      enumeration: [current]
+    value(NX_FLOAT):
+      unit: NX_TEMPERATURE
```
NX_CURRENT

lukaspie replied:
fixed

rettigl commented on contributed_definitions/nyaml/NXmanipulator.yaml:
```diff
@@ -13,36 +13,106 @@ NXmanipulator(NXobject):
+    value(NX_FLOAT):
+      unit: NX_TEMPERATURE
+      doc: |
+        In the case of a fixed drain current measurement, this is the scalar drain current measured between
```
"in case of a single or averaged"...

lukaspie replied:
fixed for all sensors/actuators

rettigl commented on contributed_definitions/nyaml/NXmanipulator.yaml:
```diff
@@ -13,36 +13,106 @@ NXmanipulator(NXobject):
+        doc: |
+          In the case of an experiment in which the current changes and is recorded with
+          time stamps, this is an array of length m of currents.
+  sample_bias_potentiostat(NXsensor):
```
NXactuator

lukaspie replied:
fixed

rettigl commented on contributed_definitions/nyaml/NXmanipulator.yaml:
```diff
@@ -13,36 +13,106 @@ NXmanipulator(NXobject):
+    measurement:
+      enumeration: [voltage]
+    setpoint(NX_FLOAT):
+      unit: NX_TEMPERATURE
```
NX_VOLTAGE

lukaspie replied:
fixed

rettigl commented on contributed_definitions/nyaml/NXmanipulator.yaml:
```diff
@@ -13,36 +13,106 @@ NXmanipulator(NXobject):
+    measurement:
+      enumeration: [voltage]
+    value(NX_FLOAT):
+      unit: NX_TEMPERATURE
```
NX_VOLTAGE

lukaspie replied:
fixed

rettigl commented on contributed_definitions/nyaml/NXmanipulator.yaml:
```diff
@@ -13,36 +13,106 @@ NXmanipulator(NXobject):
-      NXsample if present.
-  drain_current(NX_FLOAT):
-    unit: NX_CURRENT
+      Any additional sensors on the manipulator used to control an external condition. 
```
additional actuators

lukaspie replied:
fixed

rettigl commented on contributed_definitions/nyaml/NXmanipulator.yaml:
```diff
@@ -13,36 +13,106 @@ NXmanipulator(NXobject):
   (NXpositioner):
     doc: |
       Class to describe the motors that are used in the manipulator
-  (NXfabrication):
```
Why was this removed?

lukaspie replied:
accidental, fixed. Thanks for spotting it.

rettigl commented on contributed_definitions/nyaml/NXmpes.yaml:
```diff
@@ -289,34 +289,92 @@ NXmpes(NXobject):
-              term: 8.41
-              url: https://www.iso.org/obp/ui/en/#iso:std:iso:18115:-1:ed-3:v1:en:term:8.41
-        device_information(NXfabrication):
+          name:
+            exists: recommended
+          acutation:
+            enumeration: [temperature]
+          type:
+            exists: optional
+          setpoint(NX_FLOAT):
+            exists: recommended
```
In this current setup, is there actually a place to e.g. store the actual heater power as function of time? This reads like a temperature setpoint, which is not the same.

lukaspie replied:
You are right, this is currently missing.

I think having a setpoint(_log) here that stores the setpoint values as a function of time is reasonable (as it is). There should then be an additional field (and its log) that stores the heater power, or similar.  I think this is why there was an additional `output` field in `NXacutator` at the beginning. However, `output` is probably not the best term. Any good suggestions? Maybe generic `PARAMETER` and `PARAMETER_log`? And then specify whenever `NXactuator `is used?

rettigl replied:
Strictly speaking, the setpoint of the sample heater as a "Temperature" requires the use of some controller that reads in the temperature and sets the heater power. So The actuator should write "heater_power", and the setpoint should be in NXpid. But not sure if that makes sense to break things up so much...

lukaspie replied:
How about this:
**NXacutator.yaml**
```yaml
NXactuator(NXobject):
  OUTPUT(NX_FLOAT):
    unit: NX_ANY
    doc: |
      Any output that the actuator produces.
      For example, a heater can have the field heater_power(NX_FLOAT).
  OUTPUT_log(NXlog):
    unit: NX_ANY
    doc: |
      Time history of actuator outputs.
  (NXpid):
    doc: |
      If the actuator is PID-controlled, the settings of the PID controller can be stored here.
    setpoint(NX_FLOAT):
      unit: NX_ANY
      doc: |
        Nominal actuator setpoint.
        Can be a scalar or a vector (of [n] actuations).
    setpoint_log(NXlog):
      doc: |
        Time history of actuator setpoints.
```

**NXpid.yaml**
```yaml
NXpid(NXobject):
  setpoint(NX_FLOAT):
    unit: NX_ANY
    doc: |
      The Setpoint(s) used as an input for the PID controller.
      
      It can also be a link to an NXsensor.value field.
  [...]
  setpoint_log(NXlog):
    unit: NX_ANY
    doc: |
      Time log of the setpoint(s) used as an input for the PID controller.
```
And then to be used in **NXmanipulator.yamö**:
```yaml
NXmanipulator(NXobject):
  [...]
  sample_heater(NXactuator):
    doc: Device to heat the sample.
    actuation:
      enumeration: [temperature]
    heater_power(NX_FLOAT):
      unit: NX_ANY
      doc: |
        In case of a fixed or averaged heating power, this is the scalar heater power.
        It can also be a 1D array of heater powers (without time stamps).
    heater_power_log(NXlog):
      unit: NX_ANY
       doc: |
         In the case of an experiment in which the heater power is changed and recorded with time stamps,
         this is an array of length m of temperature setpoints.
    (NXpid):
      setpoint(NX_FLOAT):
        unit: NX_TEMPERATURE
        doc: |
          In case of a fixed or averaged temperature, this is the scalar temperature setpoint.
          It can also be a 1D array of temperature setpoints (without time stamps).
      setpoint_log(NXlog):
        value(NX_NUMBER):
          unit: NX_TEMPERATURE
          doc: |
            In the case of an experiment in which the temperature is changed and the setpoints are
            recorded with time stamps, this is an array of length m of temperature setpoints.
```
@domna also curious about your thoughts.

domna replied:
> How about this: **NXacutator.yaml**
> 
> ```yaml
> NXactuator(NXobject):
>   OUTPUT(NX_FLOAT):
>     unit: NX_ANY
>     doc: |
>       Any output that the actuator produces.
>       For example, a heater can have the field heater_power(NX_FLOAT).
>   OUTPUT_log(NXlog):
>     unit: NX_ANY
>     doc: |
>       Time history of actuator outputs.
> ...
> ```

This design actually means that every field of type `NX_FLOAT` in the NXactuator class is an output, because there is now way of distinguishing this. If this is fine this is ok, but maybe it's better to have some kind of prefix? i.e., `output_OUTPUT` and `output_OUTPUT_log`?

I also thought whether `actuation` is a good vocabulary for the quantity being actuated? Maybe something like `physical_quantity` is a bit easier to understand from it's name, since it's already clear that we are dealing with something which is actuated in this group.

Other than those two points I like the general design.

lukaspie replied:
> This design actually means that every field of type NX_FLOAT in the NXactuator class is an output, because there is now way of distinguishing this. If this is fine this is ok, but maybe it's better to have some kind of prefix? i.e., output_OUTPUT and output_OUTPUT_log?

I already thought about this. `output_OUTPUT_log` seems very abstract and not easy to understand. There are currently no other field of type `NX_FLOAT` in NXactuator, so my proposal would still work. I guess if the base class is extended later, this could be problematic.

> I also thought whether actuation is a good vocabulary for the quantity being actuated? Maybe something like physical_quantity is a bit easier to understand from it's name, since it's already clear that we are dealing with something which is actuated in this group.

I chose `actuation` to mirror `measurement` in `NXsensor`. But I agree, `physical_quantity` probably makes more sense. I have renamed it.

domna replied:
> > This design actually means that every field of type NX_FLOAT in the NXactuator class is an output, because there is now way of distinguishing this. If this is fine this is ok, but maybe it's better to have some kind of prefix? i.e., output_OUTPUT and output_OUTPUT_log?
> 
> I already thought about this. `output_OUTPUT_log` seems very abstract and not easy to understand. There are currently no other field of type `NX_FLOAT` in NXactuator, so my proposal would still work. I guess if the base class is extended later, this could be problematic.

Actually, this would not be problematic for the definitions, as new fields can just be added. However, if someone wants to store a field of type `NX_FLOAT` in NXactuator which is not in this base class it would automatically be assumed that this is an `OUTPUT` field. I think this could be very confusing for nexus users.
> 
> > I also thought whether actuation is a good vocabulary for the quantity being actuated? Maybe something like physical_quantity is a bit easier to understand from it's name, since it's already clear that we are dealing with something which is actuated in this group.
> 
> I chose `actuation` to mirror `measurement` in `NXsensor`. But I agree, `physical_quantity` probably makes more sense. I have renamed it.

👍️

rettigl replied:
Should we not also change measurement in NXsensor then to physical_quantity?

rettigl replied:
Otherwise, I find this solution good, and don't see a problem with the potentially assignement of undocumented fields. It's the same in NXdata or NXtransformations, no?

domna replied:
> Otherwise, I find this solution good, and don't see a problem with the potentially assignement of undocumented fields. It's the same in NXdata or NXtransformations, no?

That's true it's handled the same there. Then let's keep it like this.

lukaspie replied:
> Should we not also change measurement in NXsensor then to physical_quantity?

This breaks the NIAC standard, so I didn't do it.

rettigl replied:
Ah, right, this is not our's...
Well then, let's keep it like it is.

rettigl commented on contributed_definitions/nyaml/NXmpes.yaml:
```diff
@@ -289,34 +289,92 @@ NXmpes(NXobject):
-              spec: ISO 18115-1:2023
-              term: 8.41
-              url: https://www.iso.org/obp/ui/en/#iso:std:iso:18115:-1:ed-3:v1:en:term:8.41
-        device_information(NXfabrication):
```
Why has the device_information been removed?

lukaspie replied:
Accidental. Has been readded to NXmanipulator.

The xref part above has been added to NXsample/bias again.

rettigl commented on contributed_definitions/nyaml/NXmpes.yaml:
```diff
@@ -289,34 +289,92 @@ NXmpes(NXobject):
+          exists: optional
+        setpoint(NX_FLOAT):
+          exists: recommended
+          unit: NX_PRESSURE
```
NX_CURRENT

lukaspie replied:
fixed

rettigl commented on contributed_definitions/nyaml/NXmpes.yaml:
```diff
@@ -289,34 +289,92 @@ NXmpes(NXobject):
+            doc: |
+              In the case of an experiment in which the gas pressure is changed and the setpoints are
+              recorded with time stamps, this is an array of length m of current setpoints.
+            unit: NX_PRESSURE
```
NX_CURRENT

rettigl commented on contributed_definitions/nyaml/NXmpes.yaml:
```diff
@@ -289,34 +289,92 @@ NXmpes(NXobject):
+          exists: optional
+          value(NX_NUMBER):
+            doc: |
+              In the case of an experiment in which the gas pressure is changed and the setpoints are
+              recorded with time stamps, this is an array of length m of current setpoints.
```
adopt

lukaspie replied:
done

rettigl reviewed (approved):
Mostly looks good, apart from a few comments and copy paste errors.

lukaspie commented:
Implemented the new structure in `NXactuator`, ready for final review.

lukaspie commented:
I just realized that one aspect of the new `NXsample `base class (i.e., using either `NXsubstance `OR `NXsample_component`) was not represented here. So, I moved `chemical_formula` to `NXsubstance `and renamed according to the base class.
